docs/markdown/Tutorials/test-custom-plugin-goal.md

```diff
@@ -0,0 +1,280 @@
+# Introduction
+
+In this tutorial, we'll learn how to test the custom plugin we wrote earlier. Pants documentation provides comprehensive coverage of the [plugin testing](https://www.pantsbuild.org/docs/rules-api-testing) and this tutorial should you to get started writing own tests.
+
+Most of the plugin code that needs to be tested is in the following files:
+* `rules.py` where we implemented how a `VERSION` file needs to be read and how to use a `version_file` BUILD target
+* `tailor.py` where we taught the `tailor` goal about the `VERSION` files and generation of `version_file` targets
+
+To get started writing tests, it may make sense to write a very high level integration test first to confirm our code does what we expect. You can think of this as of [integration tests for Pants](https://www.pantsbuild.org/docs/rules-api-testing#approach-4-run_pants-integration-tests-for-pants). Let's run our goal from a test!
+
+## Testing with a complete Pants process
+
+Pants provides a convenient way to run a full Pants process as it would run on the command line. Writing such a test would be equal to having, say, a Shell script to confirm that the output of the `./pants project-version myapp:` command is `{"path": "myapp/VERSION", "version": "0.0.1"}`. Keep in mind that running custom scripts with this type of tests would require having a Pants repository set up (including the `pants.toml` configuration), creating `BUILD` metadata files and so on. When writing custom acceptance tests using `pants.testutil` package, in contrast, you don't have to worry about that and can focus on testing your plugin logic in the very minimalistic environment containing only what's absolutely necessary to run your plugin code. 
+
+In the following code snippet, we define a set of files to be created (in a temporary directory that Pants manages for us), the backends to be used (Python and our custom plugin), and a Pants command to be run. By reading the `stdout` of a process, we can confirm the plugin works as expected (conveniently ignoring any unrelated warnings that Pants may have produced).
+
+```python
+import json
+from pathlib import Path
+
+from pants.testutil.pants_integration_test import run_pants, setup_tmpdir
+
+build_root_marker = Path.cwd().joinpath("BUILDROOT")
+
+
+def test_reading_project_version_target() -> None:
+    """Run a full Pants process as it would run on the command line."""
+    project_files = {
+        "project/BUILD": "version_file(source='VERSION')",
+        "project/VERSION": "10.6.1",
+    }
+    # This is a limitation of the current implementation.
+    # See https://github.com/pantsbuild/pants/issues/12760.
```

Comment:
Suggest having a issue linked to #12760 that instructs to clean this up once it has been fixed. Otherwise this will become stale/incorrect once the issue is fixed and we'll likely miss to update here. :)

Reply:
Great idea, created https://github.com/pantsbuild/pants/issues/18027